VMware: match nic mac for ip address fetch

alexandru-bagu opened the pull request:
### Description

This PR fixes #10640

The main solution for the issue here would be to make sure that we are getting the IP for the correct NIC which we can do based on the MAC of the NIC. Additionally, a null IP is also a valid response because a NIC can have no IP assigned.

It's up for discussion whether we should pull any IP even if it does not match the network subnet because the user can in fact use the NIC with another subnet as well, but that is not necessarily relevant to this issue.
 
<!--- ******************************************************************************* -->
<!--- NOTE: AUTOMATION USES THE DESCRIPTIONS TO SET LABELS AND PRODUCE DOCUMENTATION. -->
<!--- PLEASE PUT AN 'X' in only **ONE** box -->
<!--- ******************************************************************************* -->

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI
- [ ] test (unit or integration test code)

### Feature/Enhancement Scale or Bug Severity

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [x] Minor
- [ ] Trivial

### Screenshots (if appropriate):

### How Has This Been Tested?

I am in the process of testing on my dev environment then going to push in prod once I am happy with the results.

#### How did you try to break this feature and the system with this change?

<!-- see how your change affects other areas of the code, etc. -->

<!-- Please read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/main/CONTRIBUTING.md) document -->

weizhouapache reviewed (commented):
overall looks good

A related pr for l2 network #10431

weizhouapache commented:
@blueorangutan package

blueorangutan commented:
@weizhouapache a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

alexandru-bagu commented on server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:
```diff
@@ -788,12 +793,6 @@ protected void runInContext() {
                         }
                     }
                 } else {
-                    //previously vm has ip and nic table has ip address. After vm restart or stop/start
```
In my opinion, if we cannot know if the VM still has the IP assigned or not then we should not modify the database. In other words, we should be sure that the VM's NIC does not have an IP assigned before updating the database.

alexandru-bagu commented:
> overall looks good
> 
> A related pr for l2 network #10431

Interesting. I see your PR has more changes than this PR. Feel free to grab my code for VMware and close this PR.

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 12924

weizhouapache commented:
@blueorangutan package

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 13098

weizhouapache commented:
@alexandru-bagu 
can you rebase with 4.20 branch so it will be included into 4.20.1.0 release ?

alexandru-bagu commented:
> @alexandru-bagu can you rebase with 4.20 branch so it will be included into 4.20.1.0 release ?

I rebased the changes onto 4.20 and changed merge target branch to 4.20.

DaanHoogland commented:
@blueorangutan package

weizhouapache commented:
> > @alexandru-bagu can you rebase with 4.20 branch so it will be included into 4.20.1.0 release ?
> 
> I rebased the changes onto 4.20 and changed merge target branch to 4.20.

great, thanks @alexandru-bagu 
I will test it next week

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 13179

DaanHoogland commented:
@alexandru-bagu , i fixed some conflict eagerly on you PR. I hope you’ll forgive someday soon. Please have a check?

alexandru-bagu commented:
Daan, #10431 handles kvm. This PR properly handles IPs detected using vmtools for VMware. This bit should be merged either way.

What I leave up for discussion is the case where the `GetVmIpAddressCommand` returns false or true with a null ip. In my opinion, if the command returns false (as in there was an error) the IP should not be released - basically the code I wrote in `server/src/main/java/com/cloud/vm/UserVmManagerImpl.java`.

DaanHoogland commented:
> Daan, #10431 handles kvm. This PR properly handles IPs detected using vmtools for VMware. This bit should be merged either way.

sorry, I did not read the code well enough :|

> What I leave up for discussion is the case where the `GetVmIpAddressCommand` returns false or true with a null ip. In my opinion, if the command returns false (as in there was an error) the IP should not be released - basically the code I wrote in `server/src/main/java/com/cloud/vm/UserVmManagerImpl.java`.

@weizhouapache , Can you chime in? This is to be resolved before merging!
cc @Pearl1594 , I think this would be good to have if we resolve this question.

weizhouapache commented:
> > What I leave up for discussion is the case where the `GetVmIpAddressCommand` returns false or true with a null ip. In my opinion, if the command returns false (as in there was an error) the IP should not be released - basically the code I wrote in `server/src/main/java/com/cloud/vm/UserVmManagerImpl.java`.
> 
> @weizhouapache , Can you chime in? This is to be resolved before merging! cc @Pearl1594 , I think this would be good to have if we resolve this question.

@DaanHoogland 
I agree with @alexandru-bagu 
the new logic seems good to me

DaanHoogland commented:
@blueorangutan package

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 13364

DaanHoogland commented:
@blueorangutan test ol8 vmware-70u3

DaanHoogland commented:
@blueorangutan test ol8 vmware-80u3

blueorangutan commented:
@DaanHoogland a [SL] Trillian-Jenkins test job (ol8 mgmt + vmware-80u3) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian Build Failed (tid-13309)<b/>

alexandru-bagu commented:
Reviewing the code I observed that the IP usage was still being decremented even though no action has been done due to an error being returned by the fetcher thread. In case of errors, since we don't actually do anything, we should not make any changes to the current state.

weizhouapache commented:
@blueorangutan package

blueorangutan commented:
@weizhouapache a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ debian ✔️ suse15. SL-JID 13448

weizhouapache commented:
@blueorangutan package

blueorangutan commented:
@weizhouapache a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: ✖️ el8 ✖️ el9  ✔️ debian ✖️ suse15. SL-JID 14858

weizhouapache commented:
@blueorangutan package

blueorangutan commented:
@weizhouapache a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

DaanHoogland commented:
@blueorangutan test ol8 vmware-80u3

blueorangutan commented:
@DaanHoogland a [SL] Trillian-Jenkins test job (ol8 mgmt + vmware-80u3) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian test result (tid-14447)</b>
Environment: vmware-80u3 (x2), zone: Advanced Networking with Mgmt server ol8
Total time taken: 75649 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr10641-t14447-vmware-80u3.zip
Smoke tests completed. 138 look OK, 3 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_deploy_more_vms_than_limit_allows | `Error` | 216.19 | test_deploy_vms_in_parallel.py
test_08_upgrade_kubernetes_ha_cluster | `Failure` | 3678.92 | test_kubernetes_clusters.py
test_01_scale_vm | `Error` | 1.54 | test_scale_vm.py
test_02_scale_vm_negative_offering_disable_scaling | `Error` | 1.47 | test_scale_vm.py
test_03_scale_vm_negative_vm_disable_scaling | `Error` | 1.49 | test_scale_vm.py
test_04_scale_vm_with_user_account | `Error` | 13.13 | test_scale_vm.py
test_05_scale_vm_dont_allow_disk_offering_change | `Error` | 1.48 | test_scale_vm.py

blueorangutan commented:
<b>[SF] Trillian Build Failed (tid-14493)<b/>

weizhouapache commented:
@blueorangutan package

weizhouapache commented on plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java:
```diff
@@ -5837,11 +5837,19 @@ protected Answer execute(GetVmIpAddressCommand cmd) {
+                            result = true;
+                            details = null;
+                            for (var ipAddr : guestInfoNic.getIpAddress()) {
+                                if (cmd.getVmNetworkCidr() == null || NetUtils.isIpWithInCidrRange(ipAddr, cmd.getVmNetworkCidr())) {
```
```suggestion
                                if (NetUtils.isValidIp4(ipAddr) && (cmd.getVmNetworkCidr() == null || NetUtils.isIpWithInCidrRange(ipAddr, cmd.getVmNetworkCidr()))) {
```

blueorangutan commented:
@weizhouapache a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

weizhouapache commented:
@blueorangutan test

blueorangutan commented:
Packaging result [SF]: ✔️ el8 ✔️ el9 ✔️ el10 ✔️ debian ✔️ suse15. SL-JID 15256

blueorangutan commented:
<b>[SF] Trillian test result (tid-14505)</b>
Environment: kvm-ol8 (x2), zone: Advanced Networking with Mgmt server ol8
Total time taken: 53992 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr10641-t14505-kvm-ol8.zip
Smoke tests completed. 141 look OK, 0 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

weizhouapache commented:
verified OK on both
- shared network without services (QuickCloud with no services) and
- L2 networks

tested with ubuntu cloud image
- assigned Ip manually

<img width="1192" height="206" alt="image" src="https://github.com/user-attachments/assets/5c954641-7aa7-4546-b9b5-57fd64c7a736" />

- waited a while
<img width="1967" height="323" alt="image" src="https://github.com/user-attachments/assets/06506fef-2cc5-4e4b-9abd-e1a41d99638b" />

- reboot the vm, the assigned IP is gone, the IP field is empty
<img width="1927" height="306" alt="image" src="https://github.com/user-attachments/assets/db8cb6b6-6181-429f-a02a-0f25c8dd1bb1" />

- assigned another IP inside the vm
<img width="1242" height="413" alt="image" src="https://github.com/user-attachments/assets/3105d07c-1f92-4e52-9eb9-f476be2096d0" />

- new IP is fetched
<img width="1926" height="283" alt="image" src="https://github.com/user-attachments/assets/04d19b3d-6926-448f-bc0f-554a23e16c78" />

blueorangutan commented:
<b>[SF] Trillian test result (tid-14506)</b>
Environment: vmware-80u3 (x2), zone: Advanced Networking with Mgmt server ol8
Total time taken: 100089 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr10641-t14506-vmware-80u3.zip
Smoke tests completed. 138 look OK, 3 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_deploy_more_vms_than_limit_allows | `Error` | 216.66 | test_deploy_vms_in_parallel.py
test_01_deployVMInSharedNetwork | `Failure` | 679.90 | test_network.py
test_02_list_snapshots_with_removed_data_store | `Error` | 12.72 | test_snapshots.py
test_02_list_snapshots_with_removed_data_store | `Error` | 12.73 | test_snapshots.py

vishesh92 commented on server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:
```diff
@@ -778,8 +778,13 @@ protected void runInContext() {
                     String vmIp = answer.getDetails();
-
-                    if (NetUtils.isValidIp4(vmIp)) {
+                    if (null == vmIp) {
```
```suggestion
                    if (vmIp == null) {
```

vishesh92 commented on plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java:
```diff
@@ -5837,11 +5837,19 @@ protected Answer execute(GetVmIpAddressCommand cmd) {
+                                if (NetUtils.isValidIp4(ipAddr) && (cmd.getVmNetworkCidr() == null || NetUtils.isIpWithInCidrRange(ipAddr, cmd.getVmNetworkCidr()))) {
+                                    details = ipAddr;
+                                }
+                            }
```
```suggestion
                            }
                            break;
                        }
```

weizhouapache commented:
@blueorangutan package

blueorangutan commented:
@weizhouapache a [SL] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
@weizhouapache a [SL] Trillian-Jenkins test job (ol8 mgmt + vmware-80u3) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian test result (tid-14514)</b>
Environment: vmware-80u3 (x2), zone: Advanced Networking with Mgmt server ol8
Total time taken: 74810 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr10641-t14514-vmware-80u3.zip
Smoke tests completed. 140 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_scale_vm | `Error` | 1.59 | test_scale_vm.py
test_02_scale_vm_negative_offering_disable_scaling | `Error` | 1.40 | test_scale_vm.py
test_03_scale_vm_negative_vm_disable_scaling | `Error` | 1.41 | test_scale_vm.py
test_04_scale_vm_with_user_account | `Error` | 11.64 | test_scale_vm.py
